Usermod seven segment display reloaded - add more options so it can be used on many different projects

KrX3D opened the pull request:
# Changelog for UsermodSSDR Update

**Configurable Settings**
- Introduced preprocessor macros for configuration:
  - `umSSDR_ENABLED` – enable/disable the usermod.
  - `umSSDR_ENABLE_AUTO_BRIGHTNESS` – control auto brightness.
  - `umSSDR_BRIGHTNESS_MIN` and `umSSDR_BRIGHTNESS_MAX` – set brightness range.
  - `umSSDR_INVERT_AUTO_BRIGHTNESS` – invert the brightness mapping.
  - `umSSDR_LUX_MIN` and `umSSDR_LUX_MAX` – define the sensor lux range.
  - `umSSDR_INVERTED`, `umSSDR_COLONBLINK`, and `umSSDR_LEADING_ZERO` – adjust display settings.
  - `umSSDR_DISPLAY_MASK` and related strings (`umSSDR_HOURS`, `umSSDR_MINUTES`, `umSSDR_SECONDS`, `umSSDR_COLONS`, `umSSDR_LIGHT`, `umSSDR_DAYS`, `umSSDR_MONTHS`, `umSSDR_YEARS`) – configure the physical layout.
- The numbers array (`umSSDRNumbers`) can now be overridden via the macro `umSSDR_NUMBERS`.

**New Variables and Adjustments**
- Added new runtime variables:
  - `umSSDRInvertAutoBrightness` – controls inversion of the brightness mapping.
  - `umSSDRLuxMin` and `umSSDRLuxMax` – sensor lux thresholds.
  - `umSSDRLight` – a string to represent a Light LED in the display.
- Grouped configuration values together for clarity.

**Display Functionality Enhancements**
- In `_overlaySevenSegmentDraw()`, added a new case `'L'` to handle the Light LED, calling `_showElements()` for `umSSDRLight`.
- Modified logic in `_showElements()` for consistent handling of digit elements and improved ordering when populating the time array.

**MQTT Publishing and Connection Updates**
- Updated `_publishMQTTint_P()` and `_publishMQTTstr_P()` to check for a connected MQTT state using `WLED_MQTT_CONNECTED` (wrapped in `#ifndef WLED_DISABLE_MQTT`).
- Adjusted `onMqttConnect()` to return early if `umSSDRDisplayTime` is disabled, avoiding unnecessary MQTT subscriptions.
- Extended MQTT state updates in `_updateMQTT()` to include the new parameters (lux limits, auto brightness inversion, and Light LED).

**Brightness Adjustment and Sensor Handling**
- Refactored the `loop()` function to:
  - Combine sensor handling for both photoresistor and BH1750.
  - Introduce a new flag (`disableUmLedControl`) for external control over LED updates.
  - Implement conditional brightness mapping that inverts the brightness curve if `umSSDRInvertAutoBrightness` is set.
  - Use a unified sensor reading variable (`lux`) updated from whichever sensor is available.

**Additional Public API**
- Added the public method `disableOutputFunction(bool state)` to allow external control over LED output (disable/enable).

**JSON API and Configuration Updates**
- Enhanced `addToJsonInfo()` by:
  - Adding a nested array to indicate if the module is disabled.
  - Including new parameters such as "Auto Brightness inverted" along with existing display settings.
- Updated `_addJSONObject()` and `readFromConfig()` to incorporate new parameters (lux values, auto brightness inversion, and Light LED) into the state and configuration JSON objects.

With these modifications, the SSDR usermod becomes even more versatile, allowing it to be used on a wide variety of segment clocks and projects.

Look at the readme -> ## Additional Projects


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a configurable Seven Segment Display usermod v2 with overlay time/date display, MQTT control, customizable segment masks, colon blinking, leading-zero handling, auto-brightness, and an external API to disable LED output.

- **Documentation**
  - Major README improvements: compile-time parameter tables, clearer settings, brightness options, LED definition examples, installation and dependency notes.

- **Chores**
  - Added build metadata and a setup script to enable conditional build options when companion usermods are detected; updated internal references.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

blazoncek commented:
@derqurps please have a look.

derqurps reviewed (commented):
@KrX3D Thank you for extending the function and the documentation of the usermod.

I added a couple of small suggestions. The only thing that i don't understand, is what the option "LED-Numbers-Light" is doing or what it is for :)

Thanks again!

KrX3D commented on usermods/seven_segment_display_reloaded/usermod_seven_segment_reloaded.h:
```diff
@@ -41,43 +68,108 @@ class UsermodSSDR : public Usermod {
     //  M - 01-12 month
     //  y - 21 last two positions of year
     //  Y - 2021 year
+    //  L - Light LED
```
something like?:

```
//  L - Light LED
//  This option defines a separate LED (or set of LEDs) that can be controlled independently 
//  from the other clock display LEDs. It can be switched on or off to provide additional 
//  lighting for the display or serve as an ambient light source, without affecting the 
//  clock's visual display of time and date.

```
or

  ```
//  L - Light LED (defines a separate LED or LEDs for lighting, can be controlled independently 
          from the clock display LEDs, providing additional light without affecting time display)
```

KrX3D replied:
done

derqurps reviewed (approved):
Looks good 👍

netmindz commented:
Just needs the merge conflict fixing, please pull in the latest code from main branch into yours. You might also need to update your readme for the new usermod build configuration

KrX3D committed:
Update ReadMe

KrX3D committed:
Update SSDR

KrX3D commented:
hi, i synced my fork and readded the changes and also added at

seven_segment_display_reloaded.cpp

 static UsermodSSDR seven_segment_display_reloaded;
 REGISTER_USERMOD(seven_segment_display_reloaded);
 
 but now im not sure how compile it with the usermod, not just this but every usermod i.e. auto_save
 
 i tried it on lolin_s2_mini and i added to custom_usermods  this and the auto save usermod. Am i right that the name from the library.json need to be added to custom_usermods  ?
 
 so i added to 
 
[env:lolin_s2_mini]
custom_usermods = audioreactive seven_segment_display_reloaded auto_save
 
 but at WLED under usermod its just empty:
 
 
![image](https://github.com/user-attachments/assets/a23d7d1d-7f0a-4a9a-8bbc-99331c98cc1c)

KrX3D committed:
usermod v2

KrX3D committed:
Update readme.md

KrX3D committed:
Update readme.md

KrX3D commented:
So i got it working. Should be finished now.

The problem that i couldnt compile it right seems to be related to

https://github.com/wled/WLED/issues/4597

On my windows 10 pc it doesnt work (where the build for 0.15.X works), so i created on my unraid server a docker with VS Code and there i can compile it...

KrX3D committed:
if "SN_Photoresistor" in usermods:
fixed some problems, made  and BH1750_v2 work again from this

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.h:
```diff
@@ -0,0 +1,234 @@
+class UsermodSSDR : public Usermod {
+  private:
+    //Runtime variables.
+    unsigned long umSSDRLastRefresh = 0;
+    unsigned long umSSDRResfreshTime = 3000;
+    uint16_t umSSDRLength = 0;
```
done

KrX3D committed:
coderabbitai improvments

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.h:
```diff
@@ -0,0 +1,236 @@
+		\/       \/
+		  <  D  >
+	  */
+	  uint8_t umSSDRNumbers[11][7] = {
+		//  A    B    C    D    E    F    G
+		{   1,   1,   1,   1,   1,   1,   0 },  // 0
+		{   0,   1,   1,   0,   0,   0,   0 },  // 1
+		{   1,   1,   0,   1,   1,   0,   1 },  // 2
+		{   1,   1,   1,   1,   0,   0,   1 },  // 3
+		{   0,   1,   1,   0,   0,   1,   1 },  // 4
+		{   1,   0,   1,   1,   0,   1,   1 },  // 5
+		{   1,   0,   1,   1,   1,   1,   1 },  // 6
+		{   1,   1,   1,   0,   0,   0,   0 },  // 7
+		{   1,   1,   1,   1,   1,   1,   1 },  // 8
+		{   1,   1,   1,   1,   0,   1,   1 },  // 9
+		{   0,   0,   0,   0,   0,   0,   0 }   // blank
+	};
+	#else
+	  uint8_t umSSDRNumbers[11][10] = umSSDR_NUMBERS;
+	#endif
```
thats intentional, since per default the array supports the default of this usermod, and the overwrite is for another use case

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.h:
```diff
@@ -0,0 +1,236 @@
+    static const char _str_months[];
+    static const char _str_years[];
+    
+    // Optional usermod dependencies
+    #ifdef USERMOD_SN_PHOTORESISTOR
+      Usermod_SN_Photoresistor *photoResistor = nullptr;
+    #else
+      #define photoResistor nullptr;
+    #endif
+    
+    #ifdef USERMOD_BH1750
+      Usermod_BH1750* bh1750 = nullptr;
+    #else
+      #define bh1750 nullptr;
+    #endif
```
done

KrX3D committed:
fix

KrX3D committed:
fixed wrong topic, added more logging with identifier [SSDR]

KrX3D committed:
log function -> unique name

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
bri is a wled global variable, so its not defined in my code

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -0,0 +1,660 @@
+    }
+  }
+
+  void UsermodSSDR::_showElements(String *map, int timevar, bool isColon, bool removeZero) {
+    if ((map != nullptr) && (*map != nullptr) && !(*map).equals("")) {
+      int length = (timevar == 0) ? 1 : log10(timevar) + 1;
+      bool addZero = false;
+      if (length == 1) {
+        length = 2;
+        addZero = true;
+      }
+      int timeArr[length];
+
+      if (addZero) {
+        if (removeZero) {
+          timeArr[0] = timevar;
+          timeArr[1] = 10;  // blank digit
+        } else {
+          timeArr[0] = timevar;
+          timeArr[1] = 0;
+        }
+      } else {
```
done

KrX3D committed:
coderabbitai improvments

kendrick90 commented:
Are any of you guys on the WLED discord? I am noticing some issues and would be down to contrib too just not sure where to start the conversation with you all. I'm noticing some weirdness with it if you set the clock into countdown mode. It is off by one if not showing seconds. And if using 12hr format in SSDR and countdown clock 00 gets automatically reset to 12. Not sure where to track SSDR development since it's a usermod and not technically part of the main WLED repo. I'm @kendrick90 there as well.

KrX3D committed:
Merge branch 'main' into usermod_seven_segment_display_reloaded

KrX3D commented:
> Are any of you guys on the WLED discord? I am noticing some issues and would be down to contrib too just not sure where to start the conversation with you all. I'm noticing some weirdness with it if you set the WLED clock into countdown mode. It is off by one minute if not showing seconds. And if using 12hr format in SSDR and countdown clock 00 hour gets automatically reset to 12. Not sure where to track SSDR development since it's a usermod and not technically part of the main WLED repo. I'm @kendrick90 there as well.

Hi,

im not using discord.
Can you discribe more your problem and how i can reproduce it?

as i understand there are 2 things

- some problem with countdown mode (WLED function? i never used it so far). i sthe countdown off by 1 min or the clock afterwards? is it off by the amount you set the countdown?
-  a problem when setting ssdr to 12hr format. so the problem is that it shows at midnight 12 instead of 00 ? As im not using 12hr format, is it normal to show 00 at midnight? Or does it depend on i.e. which country yiu are? so what i want to know is if i should change it from 12 to 00 or make an option to be able to pick 12/00

KrX3D committed:
skip function if usermod disabled, GUI Info - UM elements now better seperated from other UM

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -0,0 +1,662 @@
+          }
+          break;
+        case 'L':
+          if (!lightDone) { // only call _showElements once
+            _showElements(&umSSDRLight, 0, 1, 0);
+            lightDone = true;
+          }
+          break;
+        _logUsermodSSDR("Unknown mask char '%c'", c);
+      }
```
done

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -0,0 +1,662 @@
+  void UsermodSSDR::_setLeds(int lednr, int lastSeenLedNr, bool range, int countSegments, int number, bool colon) {
+    if (!umSSDRMask || (lednr < 0) || (lednr >= umSSDRLength)) return;  // prevent array bounds violation
+
+    if (!(colon && umSSDRColonblink) && ((number < 0) || (countSegments < 0))) return;
+    
+    if ((colon && umSSDRColonblink) || umSSDRNumbers[number][countSegments]) {
+      if (range) {
+        for(int i = max(0, lastSeenLedNr); i <= lednr; i++) {
+          umSSDRMask[i] = true;
+        }
+      } else {
+        umSSDRMask[lednr] = true;
```
added

KrX3D committed:
coderabbitai improvements

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
done

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
done

KrX3D committed:
coderabbitai improvments

derqurps reviewed (changes requested):
With the two deletions reverted, the usermod works. Could you fix that? Other than that, it works :)

KrX3D committed:
fix for derqurps findings

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -258,23 +258,6 @@
     #endif
   }
 
-  bool UsermodSSDR::_cmpIntSetting_P(char *topic, char *payload, const char *setting, void *value)
```
Hi, seems i forgot to update the .h file. Should work now

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -424,13 +407,15 @@
         if (lux >= 0) { // Ensure we got a valid lux value
           uint16_t brightness;
-          // Constrain lux values within defined range
-          float constrainedLux = constrain(lux, umSSDRLuxMin, umSSDRLuxMax);
```
hi nice finding, it seems i removed it by accident. I added it back

KrX3D committed:
indentation fix

KrX3D committed:
coderabbitai and indentation fix

KrX3D committed:
indentation

derqurps reviewed (approved):
Thanks for adding the fixes. From my side, all checks out. 😀

KrX3D committed:
resolve conflicts with newest library.json and setup_deps.py

KrX3D committed:
Merge branch 'main' into usermod_seven_segment_display_reloaded

KrX3D committed:
Merge branch 'wled:main' into usermod_seven_segment_display_reloaded

mgerring commented:
Hi friends: I'm bumping this PR because I just spent several hours banging my head against the wall to determine why it wasn't possible to compile WLED with the SSDR Reloaded usermod enabled. I merged this pull request into my local copy and was able to build easily. It looks like this PR is passing, I recommend approving it!

KrX3D committed:
Handle SSDR external disable state

KrX3D committed:
Merge pull request #2 from KrX3D/codex/review-usermod-for-bugs-and-improvements

Handle external LED output disable in SSDR usermod

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -0,0 +1,680 @@
+
+#ifdef WLED_DISABLE_MQTT
+#error "This user mod requires MQTT to be enabled."
+#endif
```
Intentionally keep the compile-time #error when WLED_DISABLE_MQTT is set, because this usermod currently assumes MQTT support in its integration flow and has no tested MQTT-disabled code path. Removing the guard would require a full refactor/audit of all MQTT callbacks and return paths first. Until that is done, failing fast is safer than allowing potentially broken builds.

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.h:
```diff
@@ -0,0 +1,276 @@
+      Usermod_BH1750* bh1750 = nullptr;
+    #else
+      #define bh1750 nullptr
+    #endif
```
created problems in the past when the else part was removed

KrX3D committed:
Fix SSDR MQTT handling and bounds checks

KrX3D committed:
Merge branch 'usermod_seven_segment_display_reloaded' into codex/review-usermod-for-bugs-and-improvements

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
```diff
@@ -0,0 +1,680 @@
+        umSSDRMask[lednr] = true;
+      }
+    }
+  }
```
done

KrX3D commented on usermods/seven_segment_display_reloaded_v2/seven_segment_display_reloaded_v2.cpp:
fixed

KrX3D committed:
Merge pull request #20 from KrX3D/codex/review-usermod-for-bugs-and-improvements

Fix SSDR MQTT handling, bounds checks, brightness clamping and README tweaks
